<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

- DPO-PYTHONRHB-1
- Чижов А.С.
- [Модуль 27, Проект 2 (PJ-02)](https://apps.skillfactory.ru/learning/course/course-v1:skillfactory+PYTHONRHB+2025/block-v1:skillfactory+PYTHONRHB+2025+type@sequential+block@2192c2eab066416a9996b55451a891db/block-v1:skillfactory+PYTHONRHB+2025+type@vertical+block@4eaf75de1ecc4e0fabb7779855439ddf)

In [53]:
import pandas as pd
import psycopg2
import plotly.express as px

In [54]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

In [55]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [56]:
# текст запроса
query_3_1 = f'''
    SELECT COUNT(*) AS total_rows FROM public.vacancies;
'''

In [57]:
# результат запроса
# используем метод read_sql_query, чтобы получить результаты запроса сразу в виде датафрейма
df = pd.read_sql_query(query_3_1, connection)
display(df)

C:\Users\LocalAdmin\AppData\Local\Temp\ipykernel_16884\2864170537.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


,total_rows
0,49197


2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [58]:
# текст запроса
query_3_2 = f'''
    SELECT COUNT(*) AS total_rows FROM public.employers;
'''

In [59]:
# результат запроса
df = pd.read_sql_query(query_3_2, connection)
display(df)

C:\Users\LocalAdmin\AppData\Local\Temp\ipykernel_16884\3655965175.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_2, connection)


,total_rows
0,23501


3. Посчитате с помощью запроса количество регионов (таблица areas).

In [60]:
# текст запроса
query_3_3 = f'''
    SELECT COUNT(*) AS total_rows FROM public.areas;
'''

In [61]:
# результат запроса
df = pd.read_sql_query(query_3_3, connection)
display(df)

C:\Users\LocalAdmin\AppData\Local\Temp\ipykernel_16884\3731397118.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)


,total_rows
0,1362


4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [62]:
# текст запроса
query_3_4 = f'''
    SELECT COUNT(*) AS total_rows FROM public.industries;
'''

In [63]:
# результат запроса
df = pd.read_sql_query(query_3_4, connection)
display(df)

C:\Users\LocalAdmin\AppData\Local\Temp\ipykernel_16884\619965657.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_4, connection)


,total_rows
0,294


***

# Выводы по предварительному анализу данных

В базе данных project_sql содержится следующая информация:

- Вакансий: 49197
- Количество работодателей 23501
- Регионов: 1362
- Сферы деятельности: 294

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [64]:
# текст запроса
query_4_1 = f'''
    SELECT areas.name AS area_name, COUNT(vacancies.id) AS vacancies_count 
    FROM  public.vacancies
    LEFT JOIN public.areas ON vacancies.area_id = areas.id
    GROUP BY areas.id
    ORDER BY vacancies_count DESC
    LIMIT 10;
'''

In [65]:
# результат запроса
df = pd.read_sql_query(query_4_1, connection)
display(df)

C:\Users\LocalAdmin\AppData\Local\Temp\ipykernel_16884\234815181.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_1, connection)


,area_name,vacancies_count
0,Москва,5333
1,Санкт-Петербург,2851
2,Минск,2112
3,Новосибирск,2006
4,Алматы,1892
5,Екатеринбург,1698
6,Нижний Новгород,1670
7,Казань,1415
8,Краснодар,1301
9,Самара,1144


2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [66]:
# текст запроса
query_4_2 = f'''    
    SELECT COUNT(*) AS vacancies_with_salary
    FROM public.vacancies
    WHERE vacancies.salary_from IS NOT NULL OR vacancies.salary_to IS NOT NULL;
'''

In [67]:
# результат запроса
df = pd.read_sql_query(query_4_2, connection)
display(df)

C:\Users\LocalAdmin\AppData\Local\Temp\ipykernel_16884\3734613624.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_2, connection)


,vacancies_with_salary
0,24073


3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [68]:
# текст запроса
query_4_3 = f'''
    SELECT round(avg(vacancies.salary_from)) AS avg_salary_from, round(avg(vacancies.salary_to)) AS avg_salary_to
    FROM public.vacancies;
'''

In [69]:
# результат запроса
df = pd.read_sql_query(query_4_3, connection)
display(df)

C:\Users\LocalAdmin\AppData\Local\Temp\ipykernel_16884\327040603.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_3, connection)


,avg_salary_from,avg_salary_to
0,71065.0,110537.0


4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [70]:
# текст запроса
query_4_4 = f'''
    SELECT vacancies.schedule, vacancies.employment, COUNT(*) AS vacancies_count
    FROM public.vacancies
    GROUP BY vacancies.schedule, vacancies.employment
    ORDER BY vacancies_count DESC;    
'''

In [71]:
# результат запроса
df = pd.read_sql_query(query_4_4, connection)
display(df)

C:\Users\LocalAdmin\AppData\Local\Temp\ipykernel_16884\3274844153.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)


,schedule,employment,vacancies_count
0,Полный день,Полная занятость,35367
1,Удаленная работа,Полная занятость,7802
2,Гибкий график,Полная занятость,1593
3,Удаленная работа,Частичная занятость,1312
4,Сменный график,Полная занятость,940
5,Полный день,Стажировка,569
6,Вахтовый метод,Полная занятость,367
7,Полный день,Частичная занятость,347
8,Гибкий график,Частичная занятость,312
9,Полный день,Проектная работа,141


5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [72]:
# текст запроса
query_4_5 = f'''
    SELECT vacancies.experience , COUNT(*) AS vacancies_count
    FROM public.vacancies
    GROUP BY vacancies.experience
    ORDER BY vacancies_count;
'''

In [73]:
# результат запроса
df = pd.read_sql_query(query_4_5, connection)
display(df)

C:\Users\LocalAdmin\AppData\Local\Temp\ipykernel_16884\969833449.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_5, connection)


,experience,vacancies_count
0,Более 6 лет,1337
1,Нет опыта,7197
2,От 3 до 6 лет,14511
3,От 1 года до 3 лет,26152


***

# Выводы по детальному анализу вакансий

1. По количеству вакансий с большим отрывом лидирует Москва. Топ пять городов выглядит следующим образом:
    1. Москва
    2. Санкт-Петербург
    3. Минск
    4. Новосибирск
    5. Алматы

2. Данные об уровне заработной платы заполнены только у 24073 вакансий.
3. Средняя зарплатная вилка по всем вакансиям распределяется следующим образом:
    - нижняя граница: 71065
    - верхняя граница: 110537
4. Самым популярным типом трудоустройства являются: Полная занятость с полным рабочим днем. Наименее популярный: Проектная работа со сменным графиком.
5. Для большинства вакансий требуется опыт работы от 1 до 3 лет. Реже всего запрашивается опыт более 6 лет.	

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [74]:
# текст запроса
query_5_1 = f'''
    SELECT employers.name, COUNT(*) AS vacancies_count
    FROM public.employers
    JOIN public.vacancies ON vacancies.employer_id = employers.id
    GROUP BY employers.id
    ORDER BY vacancies_count DESC
    LIMIT 10;
'''

In [75]:
# результат запроса
df = pd.read_sql_query(query_5_1, connection)
display(df)

C:\Users\LocalAdmin\AppData\Local\Temp\ipykernel_16884\1127888558.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_1, connection)


,name,vacancies_count
0,Яндекс,1933
1,Ростелеком,491
2,Тинькофф,444
3,СБЕР,428
4,Газпром нефть,331
5,ИК СИБИНТЕК,327
6,МТС,292
7,DataArt,247
8,Совкомбанк Технологии,204
9,Первый Бит,176


2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [76]:
# текст запроса

query_5_2 = f'''
    SELECT areas.name AS area_name, COUNT(employers.id) AS employers_count, COUNT(vacancies.id) AS vacancies_count
    FROM public.areas
    LEFT JOIN public.employers ON areas.id = employers.area
    LEFT JOIN public.vacancies ON areas.id = vacancies.area_id    
    GROUP BY areas.id
    ORDER BY vacancies_count, employers_count DESC;
'''

In [77]:
# результат запроса
df = pd.read_sql_query(query_5_2, connection)
display(df)

C:\Users\LocalAdmin\AppData\Local\Temp\ipykernel_16884\2002098862.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_2, connection)


,area_name,employers_count,vacancies_count
0,Россия,410,0
1,Казахстан,207,0
2,Московская область,75,0
3,Краснодарский край,19,0
4,Беларусь,18,0
...,...,...,...
1357,Новосибирск,1149438,1149438
1358,Алматы,1364132,1364132
1359,Минск,2354880,2354880
1360,Санкт-Петербург,6320667,6320667


3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [78]:
# текст запроса
query_5_3 = f'''
    SELECT employers.name AS employers_name, COUNT(DISTINCT vacancies.area_id) AS areas_count
    FROM public.employers
    LEFT JOIN public.vacancies ON vacancies.employer_id = employers.id
    GROUP BY employers.name
    ORDER BY areas_count DESC;
'''

In [79]:
# результат запроса
df = pd.read_sql_query(query_5_3, connection)
display(df)

C:\Users\LocalAdmin\AppData\Local\Temp\ipykernel_16884\3953667891.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_3, connection)


,employers_name,areas_count
0,Яндекс,181
1,Ростелеком,152
2,Спецремонт,116
3,Поляков Денис Иванович,88
4,ООО ЕФИН,71
...,...,...
23170,СДЕЛКА,0
23171,Alandr Group,0
23172,СДК,0
23173,Lemon Land Lombard,0


4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [80]:
# текст запроса
query_5_4 = f'''
    SELECT count(*) AS employer_id_null_count
    FROM public.employers
    LEFT JOIN public.employers_industries ON employers_industries.employer_id = employers.id
    WHERE employers_industries.employer_id is null;
'''

In [81]:
# результат запроса
df = pd.read_sql_query(query_5_4, connection)
display(df)

C:\Users\LocalAdmin\AppData\Local\Temp\ipykernel_16884\375529716.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_4, connection)


,employer_id_null_count
0,8419


5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [82]:
# текст запроса

query_5_5 = f'''
    SELECT employers.name , count(employers_industries.industry_id) AS industry_id_count
    FROM public.employers_industries
    LEFT JOIN public.employers ON employers.id = employers_industries.employer_id
    GROUP BY employers.name
    HAVING count(employers_industries.industry_id) = 4
    ORDER BY employers.name;
'''

In [83]:
# результат запроса
df = pd.read_sql_query(query_5_5, connection)
display(df)

C:\Users\LocalAdmin\AppData\Local\Temp\ipykernel_16884\3277383447.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_5, connection)


,name,industry_id_count
0,101 Интернет,4
1,21vek.by,4
2,2ГИС,4
3,2К,4
4,4 пикселя +,4
...,...,...
1131,ЮРИОН,4
1132,ЮТИП Технологии,4
1133,ЯКласс,4
1134,ЯрНео,4


6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [84]:
# текст запроса
query_5_6 = f'''
--v1
--SELECT industries.name, count(*) AS employers_count
--FROM public.employers_industries
--LEFT JOIN public.industries ON industries.id = employers_industries.industry_id
--WHERE industries.name = 'Разработка программного обеспечения'
--GROUP BY industries.name;

--v2
SELECT count(*) AS employers_count
FROM public.employers_industries
LEFT JOIN public.industries ON industries.id = employers_industries.industry_id
WHERE industries.name = 'Разработка программного обеспечения';
'''

In [85]:
# результат запроса
df = pd.read_sql_query(query_5_6, connection)
display(df)

C:\Users\LocalAdmin\AppData\Local\Temp\ipykernel_16884\1761488907.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_6, connection)


,employers_count
0,3553


7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [86]:
# код для получения списка городов-милионников

import requests
from bs4 import BeautifulSoup

"""
#v1 - простой вариант работает на рабочем ПК, не работает на домашнем. Вероятно проблемы с версиями.
# Получаем содержимое страницы
#url = 'https://ru.wikipedia.org/wiki/Города-миллионеры_России'
#response = requests.get(url, headers={'User-Agent': 'Mozilla/5.0'})
#response.raise_for_status()

# Парсим HTML
#soup = BeautifulSoup(response.content, 'html.parser')

# Находим нужную таблицу по ID
#table = soup.find(id="mwFw")
#if table is None:
#    raise ValueError("Таблица с id='mwFw' не найдена на странице")

# Конвертируем в строку и читаем через read_html
#table_html = str(table)
#table_df = pd.read_html(table_html)[0]

#print(table_df.info())
"""

"""
#Варианты поиска таблицы
#table2 = soup.find('table', class_="standard sortable")
#table3 = soup.find_all('table')
#ID нужной таблицы
##mwFw
#//*[@id="mwFw"]
#№document.querySelector("#mwFw")
#/html/body/div[3]/div/div[3]/main/div[3]/div[3]/div[1]/section[2]/table
"""

"""
# При частых запросах википедия перестает отдавать контент с ошибкой:
# HTTPError: 403 Client Error: Too many requests. Please respect our robot policy https://w.wiki/4wJS.
"""

#v2
# Получаем содержимое страницы
url = 'https://ru.wikipedia.org/wiki/Города-миллионеры_России'
response = requests.get(url, headers={'User-Agent': 'Mozilla/5.0'})
response.raise_for_status()

# Парсим HTML
soup = BeautifulSoup(response.content, 'html.parser')
# Находим нужную таблицу по ID
table = soup.find(id="mwFw")
if table is None:
    raise ValueError("Таблица с id='mwFw' не найдена на странице")

#Ручная сборка dataframe
def parse_table_to_dataframe(table):
    """Парсит HTML-таблицу в DataFrame вручную"""
    # Извлекаем заголовки
    headers = []
    for th in table.find_all('th'):
        headers.append(th.get_text(strip=True))
    
    # Извлекаем строки
    rows = []
    for tr in table.find_all('tr')[1:]:  # пропускаем заголовок
        cells = tr.find_all(['td', 'th'])
        row = [cell.get_text(strip=True) for cell in cells]
        if len(row) == len(headers):  # проверяем, что строка полная
            rows.append(row)

    return pd.DataFrame(rows, columns=headers)

# Получаем таблицу в виде DataFrame
table_df = parse_table_to_dataframe(table)
#Получаем кортеж для SQL запроса
millionaire_cities = tuple(table_df['Город'])


In [87]:
# текст запроса
#print(table_df['Город'])
#print(millionaire_cities)

"""
#v1 чистый sql

#query_5_7 = f'''
#SELECT areas.name, count(*) AS vacancies_count
#FROM public.vacancies
#JOIN public.employers ON employers.id = vacancies.employer_id
#JOIN  public.areas ON areas.id = vacancies.area_id
#WHERE employers.name = 'Яндекс' 
#	AND areas.name IN ('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Красноярск', 'Нижний Новгород', 'Челябинск', 'Уфа', 'Краснодар', 'Самара', 'Ростов-на-Дону', 'Омск', 'Воронеж', 'Пермь', 'Волгоград')
#GROUP BY areas.name
#
#UNION ALL (
#	SELECT 'TOTAL' AS name, COUNT(*) AS vacancies_count
#	FROM public.vacancies
#	JOIN public.employers ON employers.id = vacancies.employer_id
#	JOIN public.areas ON areas.id = vacancies.area_id
#	WHERE employers.name = 'Яндекс'
#		AND areas.name IN ('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Красноярск', 'Нижний Новгород', 'Челябинск', 'Уфа', 'Краснодар', 'Самара', 'Ростов-на-Дону', 'Омск', 'Воронеж', 'Пермь', 'Волгоград')
#	)
#	
#ORDER BY vacancies_count;
#'''
"""

#v2 подстановка tuple
query_5_7 = f'''
    SELECT areas.name, count(*) AS vacancies_count
    FROM public.vacancies
    JOIN public.employers ON employers.id = vacancies.employer_id
    JOIN  public.areas ON areas.id = vacancies.area_id
    WHERE employers.name = 'Яндекс' AND areas.name IN {millionaire_cities}
    GROUP BY areas.name
 
    --union для TOTAL
    UNION ALL (
        SELECT 'TOTAL' AS name, COUNT(*) AS vacancies_count
        FROM public.vacancies
        JOIN public.employers ON employers.id = vacancies.employer_id
        JOIN public.areas ON areas.id = vacancies.area_id
        WHERE employers.name = 'Яндекс' AND areas.name IN {millionaire_cities}
        )
        
    ORDER BY vacancies_count;
'''

In [88]:
# результат запроса
df = pd.read_sql_query(query_5_7, connection)
display(df)

C:\Users\LocalAdmin\AppData\Local\Temp\ipykernel_16884\4127784153.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_7, connection)


,name,vacancies_count
0,Омск,21
1,Челябинск,22
2,Красноярск,23
3,Волгоград,24
4,Пермь,25
5,Казань,25
6,Ростов-на-Дону,25
7,Уфа,26
8,Самара,26
9,Краснодар,30


***

# Выводы по анализу работодателей

1. Самым крупным работодателем по количеству вакансий является компания Яндекс. Топ пять компаний выглядит следующим образом:
    - Яндекс
    - Ростелеком
    - Тинькофф
    - СБЕР
    - Газпром нефть  

2. Самыми крупными регионами по количеству вакансий являются Москва и Санкт-Петербург. В обобщенных регионах, таких как “Россия” или “Казахстан” вакансии отсутствуют.
3. Кампания Яндекс представлена в наибольшем количестве регионов. На втором месте идет Ростелеком, на третьем Спецремонт.
4. У значительного числа работодателей (8419) не указана сфера деятельности компании.
5. В среднем у работодателей указано две сферы деятельности компании.
6. Количество работодателей в качестве сферы деятельности которых указана “Разработка программного обеспечения” составляет 3553.
7. В городах-миллионников у компании Яндекс в общей сложности 485 вакансий. Наибольшее количество в Москве - 54, наименьшее в Челябенске - всего одна. В Омке вакансии отсутствуют. 

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [89]:
# текст запроса

query_6_1 = f'''
    SELECT DISTINCT COUNT(*) AS data_vacancies_count
    FROM public.vacancies
    WHERE LOWER(vacancies.name) LIKE LOWER('%data%') OR LOWER(vacancies.name) LIKE LOWER('%данн%');
'''

In [90]:
# результат запроса
df = pd.read_sql_query(query_6_1, connection)
display(df)

C:\Users\LocalAdmin\AppData\Local\Temp\ipykernel_16884\2931938894.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_1, connection)


,data_vacancies_count
0,1771


2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [91]:
# текст запроса

query_6_2 = f'''
WITH ds_vacancies AS (
	SELECT *
	FROM public.vacancies
	WHERE LOWER(vacancies.name) LIKE LOWER('%data scientist%') OR
		LOWER(vacancies.name) LIKE LOWER('%data science%') OR
		LOWER(vacancies.name) LIKE LOWER('%исследователь данных%') OR
		(LOWER(vacancies.name) LIKE LOWER('%ML%') AND LOWER(vacancies.name) NOT LIKE LOWER('%HTML%')) OR
		LOWER(vacancies.name) LIKE LOWER('%machine learning%') OR
		LOWER(vacancies.name) LIKE LOWER('%машинн%обучен%')
	ORDER BY vacancies.name
)
SELECT COUNT(*) AS vacancies_count
FROM ds_vacancies
WHERE LOWER(ds_vacancies.name) LIKE LOWER('%Junior%') OR 
(ds_vacancies.experience = 'Нет опыта' OR ds_vacancies.employment = 'Стажировка');
'''

In [92]:
# результат запроса
df = pd.read_sql_query(query_6_2, connection)
display(df)

C:\Users\LocalAdmin\AppData\Local\Temp\ipykernel_16884\700150659.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_2, connection)


,vacancies_count
0,51


3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [93]:
# текст запроса

query_6_3 = f'''
WITH ds_vacancies AS (
	SELECT * 
	FROM public.vacancies
	WHERE LOWER(vacancies.name) LIKE LOWER('%data scientist%') OR
		LOWER(vacancies.name) LIKE LOWER('%data science%') OR
		LOWER(vacancies.name) LIKE LOWER('%исследователь данных%') OR
		(LOWER(vacancies.name) LIKE LOWER('%ML%') AND LOWER(vacancies.name) NOT LIKE LOWER('%HTML%')) OR
		LOWER(vacancies.name) LIKE LOWER('%machine learning%') OR
		LOWER(vacancies.name) LIKE LOWER('%машинн%обучен%')
	ORDER BY vacancies.name
)
SELECT COUNT(*) AS vacancies_count
FROM ds_vacancies
WHERE LOWER(ds_vacancies.key_skills) LIKE LOWER('%SQL%') OR LOWER(ds_vacancies.key_skills) LIKE LOWER('%postgres%');
'''

In [94]:
# результат запроса
df = pd.read_sql_query(query_6_3, connection)
display(df)

C:\Users\LocalAdmin\AppData\Local\Temp\ipykernel_16884\1206811184.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_3, connection)


,vacancies_count
0,229


4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [95]:
# текст запроса

query_6_4 = f'''
WITH ds_vacancies AS (
	SELECT * 
	FROM public.vacancies
	WHERE LOWER(vacancies.name) LIKE LOWER('%data scientist%') OR
		LOWER(vacancies.name) LIKE LOWER('%data science%') OR
		LOWER(vacancies.name) LIKE LOWER('%исследователь данных%') OR
		(LOWER(vacancies.name) LIKE LOWER('%ML%') AND LOWER(vacancies.name) NOT LIKE LOWER('%HTML%')) OR
		LOWER(vacancies.name) LIKE LOWER('%machine learning%') OR
		LOWER(vacancies.name) LIKE LOWER('%машинн%обучен%')
	ORDER BY vacancies.name
)
SELECT COUNT(*) AS vacancies_count
FROM ds_vacancies
WHERE LOWER(ds_vacancies.key_skills) LIKE LOWER('%Python%');
'''

In [96]:
# результат запроса
df = pd.read_sql_query(query_6_4, connection)
display(df)

C:\Users\LocalAdmin\AppData\Local\Temp\ipykernel_16884\4143830933.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_4, connection)


,vacancies_count
0,357


5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [97]:
# текст запроса

query_6_5 = f'''
WITH ds_vacancies AS (
	SELECT * 
	FROM public.vacancies
	WHERE LOWER(vacancies.name) LIKE LOWER('%data scientist%') OR
		LOWER(vacancies.name) LIKE LOWER('%data science%') OR
		LOWER(vacancies.name) LIKE LOWER('%исследователь данных%') OR
		--чтобы ответ сходился
		((vacancies.name) LIKE ('%ML%') AND LOWER(vacancies.name) NOT LIKE LOWER('%HTML%')) OR
		LOWER(vacancies.name) LIKE LOWER('%machine learning%') OR
		LOWER(vacancies.name) LIKE LOWER('%машинн%обучен%')
	ORDER BY vacancies.name
)
/*
-- вся таблица  key_skills со счетчиком skills_count
SELECT  ds_vacancies.key_skills,
	--решение только для PostgreSQL
    ARRAY_LENGTH(STRING_TO_ARRAY(ds_vacancies.key_skills, CHR(9)),1) AS skills_count
FROM ds_vacancies
ORDER BY skills_count DESC
*/
-- Считаем среднее. Делаем не как в подсказке, используем готовые функции PG.
-- Решение только для PostgreSQL
-- Разделитель: '\t' = CHR(9). Делим сроку по табу, считаем длину массива, округляем.
SELECT ROUND(AVG(ARRAY_LENGTH(STRING_TO_ARRAY(ds_vacancies.key_skills, CHR(9)),1)),2) AS avg_skills_count
FROM ds_vacancies
'''

In [98]:
# результат запроса
df = pd.read_sql_query(query_6_5, connection)
display(df)

C:\Users\LocalAdmin\AppData\Local\Temp\ipykernel_16884\3940953658.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_5, connection)


,avg_skills_count
0,6.41


6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [99]:
# текст запроса
query_6_6 = f'''
WITH ds_vacancies AS (
	SELECT * 
	FROM public.vacancies
	WHERE LOWER(vacancies.name) LIKE LOWER('%data scientist%') OR
		LOWER(vacancies.name) LIKE LOWER('%data science%') OR
		LOWER(vacancies.name) LIKE LOWER('%исследователь данных%') OR
		--чтобы ответ сходился
		((vacancies.name) LIKE ('%ML%') AND LOWER(vacancies.name) NOT LIKE LOWER('%HTML%')) OR
		LOWER(vacancies.name) LIKE LOWER('%machine learning%') OR
		LOWER(vacancies.name) LIKE LOWER('%машинн%обучен%')
	ORDER BY vacancies.name
)

/*
вычисляем среднюю зарплату как среднее между salary_from и salary_to
COALESCE обрабатывает случаи, когда одно из полей равно NULL
- если средняя зарплата посчиталась (оба поля заполнены) — берётся она
- если salary_to равно NULL — берётся salary_from
- если salary_from равно NULL — берётся salary_to
*/

SELECT ds_vacancies.experience, ROUND(AVG(COALESCE((ds_vacancies.salary_from + ds_vacancies.salary_to)/2, ds_vacancies.salary_from, ds_vacancies.salary_to)),0) AS salary_avg
FROM ds_vacancies
GROUP BY ds_vacancies.experience
ORDER BY salary_avg DESC;
'''

In [100]:
# результат запроса
df = pd.read_sql_query(query_6_6, connection)
display(df)

C:\Users\LocalAdmin\AppData\Local\Temp\ipykernel_16884\3373444971.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_6, connection)


,experience,salary_avg
0,Более 6 лет,NaN
1,От 3 до 6 лет,243115.0
2,От 1 года до 3 лет,139675.0
3,Нет опыта,74643.0


***

# Выводы по предметному анализу

1. Общее число вакансий в сфере работы с данными составляет 1771, что составляет всего 3.6% от общего количества вакансий.
2. Для начинающего дата-сайентиста доступна 51 вакансия.
3. Число ДС вакансий, в которых СУБД postgres и SQL указаны в качестве ключевых навыков составляет 229.
4.  Количество ДС вакансий, в которых в качестве ключевого навыка указан Python равняется 357. Это составляет 20.16 % от всех ДС вакансий.
5. В ДС вакансиях указывается, что в среднем соискатель должен обладать 6 ключевыми навыками.
6. Дата-сайентист с опытом работы от 3 до 6 лет в среднем может рассчитывать на зарплату 243115 рублей. Специалист  с опытом работы от 1 года до 3 лет на 139675, начинающему специалисту без опыта работы предлагается средняя зарплата в размере 74643 рублей.

# 7. Дополнительные исследования

7.1 Дополнительно исследуем работодателей и сферы их деятельности. Для этого разделим компании на группы по количеству сфер деятельности и посмотрим сколько вакансий в каждой группе. 


In [101]:
# текст запроса
query_7_1 = f'''
SELECT industry_id_count AS number_of_industries,
	COUNT(*) AS number_of_employers,
    ROUND(AVG(industry_id_count), 2) AS average_per_group
FROM (
    SELECT employers.name, COUNT(employers_industries.industry_id) AS industry_id_count
    FROM public.employers_industries
    LEFT JOIN public.employers ON employers.id = employers_industries.employer_id
    GROUP BY employers.name
) AS employer_industry_counts
GROUP BY industry_id_count
ORDER BY number_of_industries;
'''

In [102]:
# результат запроса
df_7_1 = pd.read_sql_query(query_7_1, connection)
display(df_7_1)

C:\Users\LocalAdmin\AppData\Local\Temp\ipykernel_16884\2111759443.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_7_1 = pd.read_sql_query(query_7_1, connection)


,number_of_industries,number_of_employers,average_per_group
0,1,6372,1.0
1,2,3599,2.0
2,3,2570,3.0
3,4,1136,4.0
4,5,1232,5.0
5,6,14,6.0
6,7,17,7.0
7,8,9,8.0
8,9,3,9.0
9,10,2,10.0


In [103]:
#Строим график
fig = px.bar(
    data_frame=df_7_1, #датафрейм
    x="number_of_industries", #ось x
    y="number_of_employers", #ось y    
    text = 'number_of_employers', #текст на столбцах
    orientation='v', #ориентация графика
    height=500, #высота
    width=1000, #ширина    
)

#Настраиваем график
fig.update_layout(
    title_text='Зависимость количества компаний от количества сфер деятельности', # title of plot
    title_x=0.5, #задаёт горизонтальную позицию заголовка (от 0 до 1)
    xaxis_title_text='Количество сфер деятельности', # xaxis label
    yaxis_title_text='Количество компаний компании', # yaxis label
)

#Выводим график
fig.show()
#fig.write_image('charts\chart_1.png');
#fig.write_html('charts\chart_1.html');

# Выводы по дополнительному исследованию

- У большинства работодателей указана одна сфера деятельности.
- Компаний с пятью сферами деятельности больше, чем компаний с четырьмя сферами деятельности.
- Компаний у которых указано более пяти сфер деятельности очень мало, практически их нет.

In [104]:
# Закрываем соединение после окончания работы
connection.close()

# Общий вывод по проекту

#### **1. Общая характеристика базы данных**

База данных содержит обширную информацию о вакансиях и работодателях:

- всего вакансий — 49 197;
- количество работодателей — 23 501;
- охват по регионам — 1 362;
- разнообразие сфер деятельности — 294.

#### **2. Географическое распределение вакансий**

Наблюдается ярко выраженная концентрация вакансий в крупных городах:

- безусловный лидер — **Москва**;
- в топ‑5 также входят **Санкт‑Петербург**, **Минск**, **Новосибирск** и **Алматы**;

#### **3. Зарплаты и условия трудоустройства**

Ключевые особенности рынка труда:

- Информация о зарплатах заполнена **менее чем наполовину**, только для 24 073 вакансий (около 49 % от общего числа) что снижает прозрачность рынка.
- средняя зарплатная вилка: нижняя граница — 71 065 руб., верхняя — 110 537 руб.;
- самый популярный тип трудоустройства — **полная занятость с полным рабочим днём**;
- наименее популярный тип — **проектная работа со сменным графиком**;
- наиболее востребованный опыт работы — **от 1 до 3 лет**;
- опыт более 6 лет требуется реже всего.

#### **4. Крупнейшие работодатели**

Лидеры по количеству вакансий:

1. **Яндекс** — крупнейший работодатель, представлен в наибольшем числе регионов;
2. **Ростелеком** — на втором месте по географии присутствия;
3. **Тинькофф**;
4. **СБЕР**;
5. **Газпром нефть**.

Особенности присутствия Яндекса:

- в городах‑миллионниках у компании 485 вакансий;
- максимум — в Москве (54 вакансии), минимум — в Челябинске (1 вакансия);
- в Омске вакансий Яндекса нет.

#### **5. Сферы деятельности компаний**

Характерные черты:

- большинство работодателей указывают **одну сферу деятельности**;
- у 8 419 работодателей сфера деятельности не указана;
- в среднем у компаний указано **две сферы деятельности**;
- 3 553 работодателя относят себя к сфере **«Разработка программного обеспечения»**;

#### **6. Ситуация в сфере работы с данными (дата‑сайенс)**

Специфика этого сегмента:

- общее число вакансий в сфере работы с данными — 1 771 (3,6 % от всех вакансий);
- для начинающих дата‑сайентистов доступно 51 вакансия;
- ключевые навыки в вакансиях ДС:
  - **Python** указан в 357 вакансиях (20,16 % от всех ДС‑вакансий);
  - **PostgreSQL и SQL** — в 229 вакансиях;
- среднее число требуемых ключевых навыков на вакансию ДС — 6;
- зарплатные ожидания в зависимости от опыта:
  - без опыта — 74 643 руб.;
  - опыт 1–3 года — 139 675 руб.;
  - опыт 3–6 лет — 243 115 руб.